### PR TITLE
Implement index selection expression parsing

### PR DESCRIPTION
**Pull request by aradi**
As a side effect, a robust error passing mechanism is also introduced.

* [x] Implement index selection parser
* [x] Use parser for all relevant keywords
* [x] Document index selection syntax.
* [x] Create entensive testing

### CODE REVIEW

**Comment by aradi** [comment]
Note: GitHub CI originally failed due to missing `findloc` in GCC-7, which comes with `ubuntu-latest` (= Ubuntu 18.04 (!) on GitHub).  I have changed the Ubuntu image in the CI now to 20.4 so that we get gfortran 9, and the CI passes. But this is open for discussion...

**Comment by awvwgk** [comment]
Since the `ubuntu-latest` image is due for the transition from 18.04 to 20.04 soon we can live with such a change. Note that even the 18.04 image has several GCC versions installed (7, 8 and 9), you can use `update-alternatives` to select another GCC version in the workflow without changing the image.

```yml
    - name: Install GCC (Linux)
      if: contains(matrix.os, 'ubuntu')
      run: |
        sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-${{ env.GCC_V }} 100 \
        --slave /usr/bin/gfortran gfortran /usr/bin/gfortran-${{ env.GCC_V }} \
        --slave /usr/bin/gcov gcov /usr/bin/gcov-${{ env.GCC_V }}
      env:
        GCC_V: 9
```

**Comment by aradi** [comment]
OK, this is ready for review now. The unit testing framework helped me to catch a lot of errors in the expression parsers. Nevertheless, it is open for discussions...

**Comment by bhourahine** [comment]
(not complaining about either point, but just curious)
- Might have been a good idea to do the unit test framework as a separate PR from the indexing changes. 
- Was there some workflow change, as various trailing white spaces have been deleted.

File: doc/dftb+/manual/waveplot.tex [hunk 1]
```diff
@@ -105,49 +105,38 @@ \subsection{Options}
...
-  represent the states, which should be plotted. The first (lowest
-  lying) state in the original \dftbp{} calculation is represented by
-  "1". The order of the specified states does not matter. You can also
-  use the specification of the form \verb|from:to| to specify
-  ranges. (For more details on range specification, see the
-  \is{MovedAtoms} keyword in the \dftbp{} manual.) The actual list of
-  molecular orbitals to plot is obtained by intersecting the
-  specifications for \is{PlottedKPoints}, \is{PlottedLevels} and
-  \is{PlottedSpins}.
+\item[\is{PlottedLevels}] The list of integers specified here represent the states, which should be
```
Comment (bhourahine) [inline doc/dftb+/manual/waveplot.tex:121]:
```suggestion
\item[\is{PlottedLevels}] The list of integers specified here represent the states which should be
```

File: prog/dftb+/include/error.fypp [hunk 2]
```diff
@@ -60,3 +60,41 @@
     end if
   end if
 #:enddef
+
+
+#! Signalizes error status.
```
Comment (bhourahine) [inline prog/dftb+/include/error.fypp:65]:
```suggestion
#! Signals error status.
```

File: prog/dftb+/include/error.fypp [hunk 3]
```diff
@@ -60,3 +60,41 @@
...
+#:def SET_ERROR(status, code, message)
+  call ${status}$%setError(${code}$, ${message}$, "${_FILE_}$", ${_LINE_}$)
+#:enddef SET_ERROR
+
+
+#! Signalizes error status with error message formatting
```
Comment (bhourahine) [inline prog/dftb+/include/error.fypp:71]:
```suggestion
#! Signals error status with error message formatting
```

File: prog/dftb+/include/error.fypp [hunk 4]
```diff
@@ -60,3 +60,41 @@
...
+#:def SET_FORMATTED_ERROR(status, code, format, *messageVars)
+  block
+    character(1024) :: message
+    write(message, ${ format }$) ${ ",".join(messageVars) }$
+    $:SET_ERROR(status, code, 'trim(message)')
+  end block
+#:enddef SET_FORMATTED_ERROR
+
+
+#! Signalizes error status and returns from curren scope
```
Comment (bhourahine) [inline prog/dftb+/include/error.fypp:81]:
```suggestion
#! Signals error status and returns from current scope
```

File: prog/dftb+/include/error.fypp [hunk 5]
```diff
@@ -60,3 +60,41 @@
...
+#:def RAISE_ERROR(status, code, message)
+  $:SET_ERROR(status, code, message)
+  return
+#:enddef RAISE_ERROR
+
+
+#! Signalizes error status with error message formatting and returns from curren scope
```
Comment (bhourahine) [inline prog/dftb+/include/error.fypp:88]:
```suggestion
#! Signals error status with error message formatting and returns from current scope
```

File: prog/dftb+/include/error.fypp [hunk 6]
```diff
@@ -60,3 +60,41 @@
...
+#:def RAISE_FORMATTED_ERROR(status, code, format, *messageVars)
+  $:SET_FORMATTED_ERROR(status, code, format, *messageVars)
+  return
+#:enddef
+
+
+#! Returns immediately from current scope if status signalizes an error.
```
Comment (bhourahine) [inline prog/dftb+/include/error.fypp:95]:
```suggestion
#! Returns immediately from current scope if status signals an error.
```

File: prog/dftb+/lib_common/status.F90 [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  implicit none
+
+  private
+  public :: TStatus
+
+
+  !> Represents status of an operation.
+  type :: TStatus
+
+    !> Status code. Zero signalizes no error, every other value is interpreted as error.
```
Comment (bhourahine) [inline prog/dftb+/lib_common/status.F90:21]:
```suggestion
    !> Status code. Zero signals no error, every other value is interpreted as an error.
```

File: prog/dftb+/lib_common/status.F90 [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    !> Instance.
+    class(TStatus), intent(inout) :: this
+
+    !> Error code.
+    integer, intent(in) :: code
+
+    !> Error message.
+    character(*), intent(in) :: message
+
+    !> File where error occured.
```
Comment (bhourahine) [inline prog/dftb+/lib_common/status.F90:56]:
```suggestion
    !> Source file where error occurred.
```

File: test/unittests/io/indexselection.F90 [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+    #! Failing tests
+
+    #:block TEST("selectWrongSpeciesName")
+      call getIndexSelection("Mg", fullRange, selected, status, speciesNames=speciesNames,&
+          & species=species)
+      @:ASSERT(status%hasError())
+      @:ASSERT(status%code == errors%invalidSelector)
+    #:endblock
+
+    #:block TEST("outOfBondsPlus")
```
Comment (bhourahine) [inline test/unittests/io/indexselection.F90:156]: Bounds not bonds and what is the plus ?
```suggestion
    #:block TEST("outOfBounds")
```

File: test/unittests/io/indexselection.F90 [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+      call getIndexSelection("17", fullRange, selected, status, speciesNames=speciesNames,&
+          & species=species)
+      @:ASSERT(status%hasError())
+      @:ASSERT(status%code == errors%invalidSelector)
+    #:endblock
+
+    #:block TEST("outOfBondsPlusZeroInd")
```
Comment (bhourahine) [inline test/unittests/io/indexselection.F90:163]:
```suggestion
    #:block TEST("outOfBoundsAndZeroInd")
```

File: test/unittests/io/indexselection.F90 [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+      call getIndexSelection("16", [0, nAtom - 1], selected, status, speciesNames=speciesNames,&
+          & species=species)
+      @:ASSERT(status%hasError())
+      @:ASSERT(status%code == errors%invalidSelector)
+    #:endblock
+
+    #:block TEST("outOfBondsMinus")
```
Comment (bhourahine) [inline test/unittests/io/indexselection.F90:170]:
```suggestion
    #:block TEST("outOfBoundsMinus")
```

File: test/unittests/io/indexselection.F90 [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+      call getIndexSelection("-17", fullRange, selected, status, speciesNames=speciesNames,&
+          & species=species)
+      @:ASSERT(status%hasError())
+      @:ASSERT(status%code == errors%invalidSelector)
+    #:endblock
+
+    #:block TEST("outOfBondsMinusZeroInd")
```
Comment (bhourahine) [inline test/unittests/io/indexselection.F90:177]:
```suggestion
    #:block TEST("outOfBoundsMinusZeroInd")
```

File: test/unittests/io/indexselection.F90 [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+    #:endblock
+
+    #:block TEST("negativeIndexWithoutBackward")
+      call getIndexSelection("-6:-2", fullRange, selected, status, speciesNames=speciesNames,&
+        & species=species, backwardIndexing=.false.)
+      @:ASSERT(status%hasError())
+      @:ASSERT(status%code == errors%invalidSelector)
+    #:endblock
+
+    #:block TEST("outOfSubrangeBonds")
```
Comment (bhourahine) [inline test/unittests/io/indexselection.F90:191]:
```suggestion
    #:block TEST("outOfSubrangeBounds")
```

File: prog/waveplot/initwaveplot.F90 [hunk 14]
```diff
@@ -441,20 +441,19 @@ subroutine readOptions(node, identity, nLevel, nKPoint, nSpin, occupations, tRea
...
-    call getChildValue(node, "PlottedLevels", buffer, child=field, &
-        &multiple=.true.)
-    call convRangeToInt(char(buffer), node, plottedLevels, nLevel)
+    call getChildValue(node, "PlottedLevels", buffer, child=field, multiple=.true.)
+    call getSelectedIndices(node, char(buffer), [1, nLevel], plottedLevels)
     if (tPeriodic) then
       call getChildValue(node, "PlottedKPoints", buffer, child=field, &
           &multiple=.true.)
-      call convRangeToInt(char(buffer), node, plottedKPoints, nKPoint)
+      call getSelectedIndices(node, char(buffer), [1, nKPoint], plottedKPoints)
```
Comment (bhourahine) [inline prog/waveplot/initwaveplot.F90:449]: k-points would be useful to have a stride operator (every 10th k-point for example).

File: prog/phonons/initphonons.F90 [hunk 15]
```diff
@@ -145,25 +145,25 @@ module dftbp_initphonons
   integer, parameter :: parserVersion = 4
 
   !> Version of the oldest parser, for which compatibility is maintained
-  integer, parameter :: minVersion = 4 
+  integer, parameter :: minVersion = 4
 
   !> Container type for parser related flags.
   type TParserFlags
```
Comment (bhourahine) [inline prog/phonons/initphonons.F90:151]: Might be worth converting comments in the TParserFlags type to doxygen/ford style.

File: prog/phonons/initphonons.F90 [hunk 16]
```diff
@@ -530,8 +529,8 @@ subroutine readFirstLayerAtoms(pnode, pls, npl, idxdevice, check)
       end if
 
   end subroutine readFirstLayerAtoms
-  
-   !!* Read bias information, used in Analysis and Green's function eigensolver 
+
+   !!* Read bias information, used in Analysis and Green's function eigensolver
```
Comment (bhourahine) [inline prog/phonons/initphonons.F90:533]:
```suggestion
   !> Read bias information, used in Analysis and Green's function solver
```

File: prog/phonons/initphonons.F90 [hunk 17]
```diff
@@ -953,19 +952,19 @@ subroutine readKPointsFile_help(child,text)
     end do
 
   end subroutine readKPointsFile_help
-      
+
 
   ! Read DFTB hessian.
```
Comment (bhourahine) [inline prog/phonons/initphonons.F90:957]:
```suggestion
  ! > Read DFTB hessian.
```

File: prog/phonons/initphonons.F90 [hunk 18]
```diff
@@ -953,19 +952,19 @@ subroutine readKPointsFile_help(child,text)
...
-  !The derivatives matrix must be stored as the following order: 
+  !The derivatives matrix must be stored as the following order:
```
Comment (bhourahine) [inline prog/phonons/initphonons.F90:958]:
```suggestion
  !> The derivatives matrix must be stored as the following order:
```

Comment (bhourahine) [inline prog/phonons/initphonons.F90:958]:
```suggestion
  !> The derivatives matrix must be stored as the following order:
```

File: prog/phonons/initphonons.F90 [hunk 19]
```diff
@@ -953,19 +952,19 @@ subroutine readKPointsFile_help(child,text)
...
   ! For the x y z directions of atoms 1..n
```
Comment (bhourahine) [inline prog/phonons/initphonons.F90:959]:
```suggestion
  !> For the x y z directions of atoms 1..n
```

File: prog/phonons/initphonons.F90 [hunk 20]
```diff
@@ -953,19 +952,19 @@ subroutine readKPointsFile_help(child,text)
...
-  !   d^2 E        d^2 E       d^2 E       d^2 E        d^2 E         
+  !   d^2 E        d^2 E       d^2 E       d^2 E        d^2 E
```
Comment (bhourahine) [inline prog/phonons/initphonons.F90:960]:
```suggestion
  ! >  d^2 E        d^2 E       d^2 E       d^2 E        d^2 E
```

File: prog/phonons/initphonons.F90 [hunk 21]
```diff
@@ -953,19 +952,19 @@ subroutine readKPointsFile_help(child,text)
...
   ! ---------- + --------- + --------- + ---------- + ---------- +...
```
Comment (bhourahine) [inline prog/phonons/initphonons.F90:961]:
```suggestion
  !> ---------- + --------- + --------- + ---------- + ---------- +...
```

File: prog/phonons/initphonons.F90 [hunk 22]
```diff
@@ -953,19 +952,19 @@ subroutine readKPointsFile_help(child,text)
...
-  ! dx_1 dx_1    dy_1 dx_1   dz_1 dx_1   dx_2 dx_1    dy_2 dx_1   
+  ! dx_1 dx_1    dy_1 dx_1   dz_1 dx_1   dx_2 dx_1    dy_2 dx_1
```
Comment (bhourahine) [inline prog/phonons/initphonons.F90:962]:
```suggestion
  !> dx_1 dx_1    dy_1 dx_1   dz_1 dx_1   dx_2 dx_1    dy_2 dx_1
```

**Review by bhourahine — APPROVED** [review]
Possibly might be worth thinking about a stride operator, e.g. 1:100:10, or a modulo operator (1:100%10) as another PR.
See suggestions in code.

File: doc/dftb+/manual/hsd.tex [hunk 23]
```diff
@@ -434,7 +434,69 @@ \section{Extended format}
...
+MovedAtoms = 1 2 3 4 5
+
+# Atoms within the range [1, 4] -> 1, 2, 3, 4
+MovedAtoms = 1:4
+
+# All atoms (from first to last)
+MovedAtoms = 1:-1
+
+# All atoms up to (and including) the fourth from the end
+# (= all atoms apart the last 3)
```
[the remark this answers is not in the document]
Comment (aradi) [reply doc/dftb+/manual/hsd.tex:468]: Why? ``-4`` always means the 4th from the end backwards.

[the remark this answers is not in the document]
Comment (bhourahine) [reply doc/dftb+/manual/hsd.tex:468]: Yes, it does, comment withdrawn.

**Pull request merged**